File: leet-1140.py
# rec TLE
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        if not piles : return 0

        suffixSum = [piles[-1]]
        for i in reversed(piles[:-1]):
            suffixSum.append(i + suffixSum[-1])
        suffixSum.reverse()

        def helper(i, m):
            if i + 2 * m >= len(suffixSum):
                return suffixSum[i]
            return suffixSum[i] - min(helper(i + x, max(m, x)) for x in range(1, 2 * m + 1))
        return helper(0, 1)
        
# DP
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        if not piles : return 0

        suffixSum = [piles[-1]]
        for i in reversed(piles[:-1]):
            suffixSum.append(i + suffixSum[-1])
        suffixSum.reverse()

        cache = {}
        def helper(i, m):
            if (i, m) in cache:
                return cache[(i, m)]
            if i + 2 * m >= len(suffixSum):
                return suffixSum[i]
            cache[(i, m)] = suffixSum[i] - min(helper(i + x, max(m, x)) for x in range(1, 2 * m + 1))
            return cache[(i, m)]
        return helper(0, 1)
        

'''
문제해석 자체가 넘 어렵네.. 뭔 말이여..
재귀형식의 일반 구현만 해내면 DP 적용은 간단한 편인데
일반 구현 조차 이해를 못하니 어렵더라..
suffix sum 개념도 알아야한다.
내가 취할 수 있는 모든 x 에 대해 상대방의 결과를 바탕으로 답을 구해내는데
일반 구현도 쉽지 않다.. 기초부터 다시공부해야지..
'''

File: legacy/leet-1140.java
import java.util.*;

class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;

        int[][] cache = new int[n][n + 1];
        int[] suffixSum = Arrays.copyOf(piles, n);

        for (int i = n - 2; i >= 0; i--) {
            suffixSum[i] += suffixSum[i + 1];
        }

        return rec(0, 1, n, cache, suffixSum);
    }

    private int rec(int i, int m, int n, int[][] cache, int[] suffixSum) {
        if (i + 2 * m >= n) return suffixSum[i];
        
        if (cache[i][m] != 0) return cache[i][m];

        for (int j = 1; j <= 2 * m; j++) {
            cache[i][m] = Math.max(cache[i][m], suffixSum[i] - rec(i + j, Math.max(m, j), n, cache, suffixSum));       
        }

        return cache[i][m];
    }
}


class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;

        int[][] cache = new int[n][n + 1];
        int[] suffixSum = Arrays.copyOf(piles, n);

        for (int i = n - 2; i >= 0; i--) {
            suffixSum[i] += suffixSum[i + 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            for (int j = 1; j <= n; j++) {
                if (i + 2 * j >= n) {
                    cache[i][j] = suffixSum[i];
                } else {
                    for (int k = 1; k <= 2 * j; k++) {
                        cache[i][j] = Math.max(cache[i][j], suffixSum[i] - cache[i + k][Math.max(j, k)]);
                    }
                }
            }
        }

        return cache[0][1];
    }
}